File: benchmark/2ddmrg.jl
using ITensors

function main()
  Ny = 6
  Nx = 12

  N = Nx*Ny

  sites = spinHalfSites(N;conserveQNs=false)

  lattice = squareLattice(Nx,Ny,yperiodic=false)
  #lattice = triangularLattice(Nx,Ny,yperiodic=false)

  ampo = AutoMPO(sites)
  for b in lattice
    add!(ampo,0.5,"S+",b.s1,"S-",b.s2)
    add!(ampo,0.5,"S-",b.s1,"S+",b.s2)
    add!(ampo,    "Sz",b.s1,"Sz",b.s2)
  end
  H = toMPO(ampo)

  state = InitState(sites)
  for n=1:N
    state[n] = isodd(n) ? "Up" : "Dn"
  end
  psi0 = MPS(state)

  sweeps = Sweeps(10)
  maxdim!(sweeps,10,20,100,100,200,400,800)
  cutoff!(sweeps,1E-8)
  @show sweeps

  energy,psi = dmrg(H,psi0,sweeps)

end
main()

File: benchmark/2ddmrg.cc
#include "itensor/all.h"
using namespace itensor;

int 
main()
    {
    int Ny = 6;
    int Nx = 12;
    bool yperiodic = false;

    auto N = Nx*Ny;

    auto sites = SpinHalf(N,{"ConserveQNs=",false});

    auto lattice = squareLattice(Nx,Ny,{"YPeriodic=",yperiodic});
    //auto lattice = triangularLattice(Nx,Ny,{"YPeriodic=",yperiodic});

    auto ampo = AutoMPO(sites);
    for(auto b : lattice)
        {
        ampo += 0.5,"S+",b.s1,"S-",b.s2;
        ampo += 0.5,"S-",b.s1,"S+",b.s2;
        ampo +=     "Sz",b.s1,"Sz",b.s2;
        }
    auto H = toMPO(ampo);

    auto state = InitState(sites);
    for(auto i : range1(N))
        {
        if(i%2 == 1) state.set(i,"Up");
        else         state.set(i,"Dn");
        }
    auto psi0 = MPS(state);

    auto sweeps = Sweeps(10);
    sweeps.maxdim() = 10,20,100,100,200,400,800;
    sweeps.cutoff() = 1E-8;
    sweeps.niter() = 2;
    sweeps.noise() = 1E-7,1E-8,0.0;
    println(sweeps);

    auto [energy,psi] = dmrg(H,psi0,sweeps,"Quiet");

    PrintData(inner(psi,H,psi));

    return 0;
    }
